### DISCUSSION TITLE
How to make sure codes execute before NavigationCompleted

### DISCUSSION THREAD

**Discussion by qingluwww** [MicrosoftEdge/WebView2Feedback#4000] [Q&A]
Hi, I encountered an extreme use case. At the same time, there are three calls to create different webview2 instances. I have some logic written in the NavigationStarting event handler, and I wish them to be executed completely before navigation completes. However, when I checked my log, there is an instance which loads a pdf, and my codes written in NavigationStarting event handler were executed after NavigationCompleted event is fired. (The other two instances seem fine.)

Is there a way to make sure the codes inside the event handler are executed completed before the navigation completes?

**Comment by ghost** [comment]
Hi, NavigationStarting event blocks the navigation, so it should be guaranteed that synchronous code in NavigationStarting handler will run before NavigationCompleted event. If you are seeing something else, feel free to share some sample code so we can look in more detail.

**Comment by qingluwww** [comment]
Thanks! What if what I load is a local file, will it behave differently? I have an assignment to be executed in js side and it's called in NavStarting event. I logged those accordingly, and from the image you can see the assignment happens actually after the NavCompleted. That's why I was wondering if the timing is the problem. (There are three different instances created at nearly the same time, so in the log there are also other instances. I know it does not make sense to do anything to the file, but there were also problems for the other instance with normal url previously, when the creation of instances has different sequence.)
![image](https://user-images.githubusercontent.com/108527580/207565767-fdd577b9-c8a0-428d-8030-f297792e7416.png)
